[FLINK-13955][runtime] migrate ContinuousFileReaderOperator to the mailbox execution model

rkhachatryan opened the pull request:
## What is the purpose of the change

Use mailbox execution model in `ContinuousFileReaderOperator`.
This allows to avoid explicit synchronization and eventually to remove `StreamTask.getCheckpointLock()`.

ContinuousFileReaderOperator was changed significantly to avoid
having an unnecessary now separate thread
and extra synchronization between blocking and non-blocking code (e.g.
format.reachedEnd and nextRecord)

Performance is lower by ~15% because of the overhead of enqueueing and processing each mail (instead of just loop).


## Brief change log

  - *Currently, `ContinuousFileReaderOperator` spawn a thread to process file contents and acquires checkpoiint lock to process each record*
  - *The new implementation enqueues a `Mail` to be executed by `MailboxProcessor` - one record at a time, until end of split is reached*
  - *ContinuousFileProcessingRescalingTest refactored and updated to work with the new model*

## Verifying this change

This change is already covered by existing tests, such as *ContinuousFileProcessingRescalingTest, ContinuousFileProcessingTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **yes, for ContinuousFileReaderOperator**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **yes, ContinuousFileReaderOperator state** (no user-visible changes)
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -96,332 +222,274 @@ public void initializeState(StateInitializationContext context) throws Exception
-				// we can ignore this
-			}
+			processRecord();
+		} catch (Exception e) {
```
`InterruptedException` might not handled properly. I'd probably not catch any exception here and let `StreamTask` handle stuff correctly. That should call `dispose`.

rkhachatryan replied:
Agree, except that `switchState(CLOSED)` should be still called to prevent further processing (even if currently it's guaranteed not to happen).

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -96,332 +222,274 @@ public void initializeState(StateInitializationContext context) throws Exception
-		super.close();
+	private void processRecord() throws IOException {
+		if (!state.prepareToProcessRecord(this)) {
+			return;
```
How can that happen? Should that result in some kind of exception?

rkhachatryan replied:
In `CLOSING` state when there is no more data
it enqueues one more mail and transitions to `CLOSED`.

Additional mail is needed to unblock `yield()` call in `close()`.

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -96,332 +222,274 @@ public void initializeState(StateInitializationContext context) throws Exception
+	@Override
+	public void dispose() throws Exception {
+		super.dispose();
+		checkpointedState = null;
```
On quite a few places `dispose` also `closes` resources. I have no idea about the actual contract though.
But it's probably safer to also call `format.close` etc from here.

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -96,332 +222,274 @@ public void initializeState(StateInitializationContext context) throws Exception
-				}
-			}
-		}
+		RunnableWithException[] runClose = {
```
According to @zentol , it's okay to use more Guava now, so I'd recommend to use `Closer`.

rkhachatryan replied:
Do you mean it's shaded version?

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -54,39 +62,157 @@
+			return TRANSITIONS
+					.getOrDefault(this, EnumSet.noneOf(ReaderState.class))
+					.contains(next);
+		}
```
Could be replaced by passing that information to the constructors of the Enums.
`ReaderState(boolean acceptingSplits, boolean isTerminal, ReaderState... allowedTransitions)`
Transitions may introduce cyclic dependencies, so not 100% sure if this part is indeed working.

rkhachatryan replied:
Constructors wouldn't work here exactly because of cyclic dependencies.

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -54,39 +62,157 @@
+		 */
+		OPENING { // the split was added and message to itself was enqueued to process it
+			@Override
+			public boolean prepareToProcessRecord(ContinuousFileReaderOperator<?> op) throws IOException {
```
A common approach would be to return the next state from the action function and always use switchState on the return (which would be a no-op in case the state returns itself). Then you wouldn't need to pass op in many cases (going towards a state machine that is not causing actions but just defining what is allowed).

rkhachatryan replied:
Using action function and decoupling transitions from states has its advantages, but it also makes the logic less obvious (if I understood you right).
Operator instance should be then passed to these transitions, so no big difference here.

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -54,39 +62,157 @@
+		IDLE {
+			@Override
+			public boolean prepareToProcessRecord(ContinuousFileReaderOperator<?> op) {
+				LOG.warn("not processing any records in IDLE state");
```
Should that be an exception?

rkhachatryan commented on flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingTest.java:
```diff
@@ -41,6 +41,10 @@
 import org.apache.flink.streaming.api.operators.StreamSource;
```
It's almost the same:) However, a separate thread IS necessary currently, before the changes.

AHeise replied:
maybe it's easier this way: "to avoid the currently necessary thread?"

rkhachatryan commented on flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingTest.java:
```diff
@@ -306,6 +311,21 @@ public int compare(String o1, String o2) {
+		CompletableFuture<OneInputStreamOperatorTestHarness<TimestampedFileInputSplit, String>> future = new CompletableFuture<>();
+		new Thread(() -> {
+			try {
+				OneInputStreamOperatorTestHarness<TimestampedFileInputSplit, String> harness = new OneInputStreamOperatorTestHarness<>(reader);
```
`Mailbox` loop must be executed by the same thread that created the `Harness` (because of thread check in the mailbox).
So only the whole method (create + run) can be moved to harness (and only to `OneInputStreamOperatorTestHarness` therefore), which I think is too specific.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -54,39 +62,157 @@
- * thread read the actual data of the split. This architecture allows the separation of the
- * reading thread from the one emitting the checkpoint barriers, thus removing any potential
- * back-pressure.
+ * <p>This implementation uses {@link MailboxExecutor} to execute each action and states:<ol>
```
fixed, thanks :)

rkhachatryan commented on flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingTest.java:
```diff
@@ -166,21 +171,21 @@ public void testFileReadingOperatorWithIngestionTime() throws Exception {
 
 		ConcurrentLinkedQueue<Object> output = tester.getOutput();
 
+		Thread.sleep(1000);
```
This is to wait for input to be processed.
Replacing with
```
while (output.isEmpty()) {
     Thread.sleep(50);
}
```
Latches would work, but make the test and harness too complex.
I think that the ideal solution would be to introduce some form of listener/controller into the Operator. But I think that with FLIP-27 it won't be necessary.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -54,39 +62,157 @@
+	/**
+	 * MUST only be changed via {@link #switchState(ReaderState) switchState}.
+	 */
+	private PriorityQueue<TimestampedFileInputSplit> splits;
```
The field should have been `transient` (fixed), and that's why it's initialized in `open()`.
Thanks for pointing out.

rkhachatryan commented:
Thanks for the feedback.

They (test changes) just went to the last commit. Fixed that.
Also updated the commit message.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -67,361 +89,422 @@
+	private transient ReaderState state = ReaderState.IDLE;
+	private transient PriorityQueue<TimestampedFileInputSplit> splits;
+	private transient TimestampedFileInputSplit currentSplit; // can't work just on queue tail because it can change because it's PQ
+	private transient Counter completedSplitsCounter;
```
There is no need for `transient` fields anymore with the factor. All of those could and should be `final` not nullable fields?

rkhachatryan replied:
`final`: the fields are nulled out in `dispose()` method (as in the previous version; not sure how important it is). And some of them can't be initialized in the constructor, because they use parent (`AbstractStreamOperator`) `transient` fields. 

`transient`: you're right, but I think it makes sense to keep `transient` modifier consistent at least within a hierarchy

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -67,361 +89,422 @@
+
+		this.reusedRecord = serializer.createInstance();
+		this.completedSplitsCounter = getMetricGroup().counter("numSplitsProcessed");
+		this.executor = executor != null ? executor : getContainingTask().getMailboxExecutorFactory().createExecutor(getOperatorConfig().getChainIndex());
```
? Mailbox executor should be already constructed from `ContinuousFileReaderOperatorFactory`

rkhachatryan replied:
It can be null when constructed from tests.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -67,361 +89,422 @@
+	private transient TimestampedFileInputSplit currentSplit; // can't work just on queue tail because it can change because it's PQ
+	private transient Counter completedSplitsCounter;
+
+	private final transient RunnableWithException runnable = () -> {
```
`s/runnable/processRecordAction/g` or `processRecordMail` or `processRecordRunnable`.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -67,361 +89,422 @@
+		Preconditions.checkState(state.isAcceptingSplits());
+		splits.offer(element.getValue());
+		if (state == ReaderState.IDLE) {
+			enqueueMail();
```
this call happens already in the mailbox thread. Is there a reason why it has to be enqueued to the mailbox queue?

rkhachatryan replied:
For each record in a file split, a mail has to be enqueued to allow process other mails.
After receiving a split we **could** process its 1st record right away. We'd need to check if there are other splits being processed and if this split is empty. 
But the potential speedup is negligible (just 1st record), so I'd rather avoid even a single extra if.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -67,361 +89,422 @@
-	@Override
-	public void processWatermark(Watermark mark) throws Exception {
-		// we do nothing because we emit our own watermarks if needed.
+	private void enqueueMail() {
```
`reenqueueProcessRecord`? `enqueueProcessRecord`

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -67,361 +89,422 @@
+				() -> output.close(),
+				() -> format.close(),
+				() -> format.closeInputFormat()};
+		Exception ef = null;
```
`ef` -> `firstException`/`originalException`.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:
```diff
@@ -67,361 +89,422 @@
 
 		checkState(checkpointedState != null,
-			"The operator state has not been properly initialized.");
+				"The operator state has not been properly initialized.");
```
nit: 
1. incorrect formatting (and in couple of lines below).
2. please try to not mix formatting (and refactoring) changes with functional changes

pnowojski commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/ContinuousFileProcessingRescalingTest.java:
```diff
@@ -112,22 +118,35 @@ private HarnessWithFormat buildAndStart(
-		if (splits != null) {
-			for (int i = 0; i < splits.length; i++) {
-				harness.processElement(new StreamRecord<>(getTimestampedSplit(i, splits[i])));
+		new Thread(() -> {
```
The same remarks as in `ContinuousFileProcessingTest`

rkhachatryan commented on flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingTest.java:
```diff
@@ -217,10 +222,11 @@ public void testFileReadingOperatorWithIngestionTime() throws Exception {
+					new TimestampedFileInputSplit(modTimes.get(split.getPath().getName()),
+							split.getSplitNumber(), split.getPath(), split.getStart(),
+							split.getLength(), split.getHostnames())));
+			tester.enqueueMail(runnableWithException);
```
If I understood you correctly, it would introduce tight coupling between the test and the implementation details (`mailbox`).
And so it won't be compatible with the upcoming FLIP-27 implementation and the test would have to be rewritten again.

rkhachatryan commented:
@pnowojski,
I've incorporated test changes you proposed, please review it.

CI failures are unrelated (azure - known issue, rebased; travis - looking into).

pnowojski commented on flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingTest.java:
```diff
@@ -313,17 +313,11 @@ public int compare(String o1, String o2) {
-			}
-		}).start();
-		return future.get();
+		return new OneInputStreamOperatorTestHarness<>(reader);
```
optional nit: inline?

pnowojski commented on flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingTest.java:
```diff
@@ -172,20 +171,21 @@ public void testFileReadingOperatorWithIngestionTime() throws Exception {
 
 		ContinuousFileReaderOperator<String> reader = new ContinuousFileReaderOperator<>(format);
 		final OneInputStreamOperatorTestHarness<TimestampedFileInputSplit, String> tester = getTester(reader);
```
missing `close()`?

rkhachatryan replied:
It is closed by `tester.close()` later.

pnowojski commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/ContinuousFileProcessingRescalingTest.java:
```diff
@@ -67,38 +65,47 @@
 	@Test
 	public void testReaderScalingDown() throws Exception {
-		HarnessWithFormat[] beforeRescale = buildAndStart(5, 15);
+		HarnessWithFormat[] beforeRescale = {};
```
optional nit: use `CloseableRegistry`?

rkhachatryan replied:
Hmm...looks like it swallows Exceptions.
I'd like to see if something wrong happens.

AHeise replied:
I recommend to use Guava's `Closer`, which will close everything and add suppressions if necessary.

pnowojski replied:
As I suggested I think using some generic `Closer` would be better, but let's not block this PR on this (optional nit). I will merge it as it is, we can clean up the test later on.

pnowojski commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/ContinuousFileProcessingRescalingTest.java:
```diff
@@ -248,26 +239,28 @@ public String nextRecord(String reuse) {
-		public void awaitFirstChunkProcessed() throws InterruptedException {
-			firstChunkTrigger.await();
+		public boolean isFirstChunkProcessed() throws InterruptedException {
+			return firstChunkTrigger.isTriggered();
```
nit: change latch to `boolean`?

rkhachatryan commented:
Thanks for the feedback.

Travis failure is unrelated (see https://issues.apache.org/jira/browse/FLINK-15951; azure and private branch passed).

pnowojski commented:
Merging
